Protocol release v1.6.0 (#45)
ElrondNetwork/elrond-specs: releases/protocol/release-specs-v1.6.0-Sirius.md

[comment]: <> (last PR processed #5490)

# Contents

This document explains the contents of the rc/v1.6.0 release codenamed Sirius. It is split in 3 sections: 
- the **features** list containing detailed insights of the feature along with the external impact and the relevant 
pull requests list
- the **smaller features and fixes** area contains the one-pull request small features or fixes along with the 
external impact details
- the **merges** area contains the list of pull requests used to update feature branches as the work progressed on 

> multiple areas at once. It is here only for reference purposes only.

Reviewer: ```suggestion
multiple areas at once. It is here for reference purposes only.
```

Author: changed

# Features

## 1. Optimise consensus signature check [#4467](https://github.com/multiversx/mx-chain-go/pull/4467)

This feature brings several changes on the way block signatures are handled during consensus in order to optimize the CPU usage.
Each signature share used in the aggregation of signatures for the block consensus proof, was previously individually verified.
Assuming all validators are honest, the verification takes ~3ms per signature. In metachain, each block accumulates ~400 signatures
which are individually checked by all participants in consensus, taking a cumulative ~1.2s CPU time for each block that passes consensus.
This 1.2 seconds for metachain and ~200 ms for shards each round (for the synchronized nodes) can be optimized out, if instead the aggregation

> is done assuming all signature shares are valid, and verification is done on the aggregation instead.

Reviewer: I would reformulate this phrase:
This 1.2 seconds required for the metachain and ~ 200 ms for the shards each round (for the synchronized nodes) can be optimized. This can be achieved by assuming all signature shares are valid during aggregation, and then performing verification on the aggregated result.

Author: changed

This feature solves a technical debt by splitting the functionality of a resolver component in 2.
The resolver component should be able only to respond to requests from the p2p network and the requester should 
be able only to request missing information (send the request packet on the p2p network)

### Impact:
* No activation epoch
* There is a small renaming in the `config.toml` file, the section `[Resolvers]` was changed to `[Requesters]`
* No binary flags changes
* No API endpoints changes

### Relevant PRs:
- [#4683](https://github.com/multiversx/mx-chain-go/pull/4683) - Created requests handler for each resolver
- [#4700](https://github.com/multiversx/mx-chain-go/pull/4700) - Added containers, factories and topic sender
- [#4705](https://github.com/multiversx/mx-chain-go/pull/4705) - Added unit tests
- [#4709](https://github.com/multiversx/mx-chain-go/pull/4709) - Renamed and moved components
- [#4715](https://github.com/multiversx/mx-chain-go/pull/4715) - Removed the requester logic out from resolvers along with multiple stubs

## 3. Multikey [#4741](https://github.com/multiversx/mx-chain-go/pull/4741)

> This feature added multikey support that will enable the node to sign on behalf of more than one key. To work, a multikey node will be required for each

Reviewer: ```suggestion
This feature added multikey support that will enable the node to sign on behalf of more than one key. For it to function, a multikey node will be required for each
```

Author: changed

- [#5057](https://github.com/multiversx/mx-chain-go/pull/5057) - Proper releases for used libraries

There is a related feature called feat/multikey metrics [#5362](https://github.com/multiversx/mx-chain-go/pull/5362) that added new API endpoint routes to provide information regarding
the managed keys status

### Relevant PRs:
- [#5356](https://github.com/multiversx/mx-chain-go/pull/5356) - Created a new component needed to provide managed keys info called managedPeersMonitor
- [#5357](https://github.com/multiversx/mx-chain-go/pull/5357) - Added new API endpoints that will provide valuable information while the node is running in multikey mode
- [#5354](https://github.com/multiversx/mx-chain-go/pull/5354) - Made the metrics update whenever one of the managed keys is leader
- [#5371](https://github.com/multiversx/mx-chain-go/pull/5371) - Fixed a member initialization that caused panic
- [#5375](https://github.com/multiversx/mx-chain-go/pull/5375) - Added `/node/managed-keys` endpoint

## 4. PubKeyConverter refactor [#4716](https://github.com/multiversx/mx-chain-go/pull/4716)

> The original implementation of the `pubPubkeyConverter.Encode` method did not propagate the error to its caller.

Reviewer: I messed up my own text :facepalm: 
```suggestion
The original implementation of the `PubkeyConverter.Encode` method did not propagate the error to its caller.
```

Author: changed

Also, the gas schedule files contain a new gas definition called `GetActiveFund`
* No binary flags changes
* No API endpoints changes

Relevant PRs:
- [#4910](https://github.com/multiversx/mx-chain-go/pull/4910) - Added new view functions for governance
- [#4913](https://github.com/multiversx/mx-chain-go/pull/4913) - Linter fixes, enabled feature in test configs
- [#4879](https://github.com/multiversx/mx-chain-go/pull/4879) - Main governance v3 branch
- [#5223](https://github.com/multiversx/mx-chain-go/pull/5223) - Added a fee when proposal is lost

## 7. DNS v2 [#5045](https://github.com/multiversx/mx-chain-go/pull/5045)

Integrated new DNS functionalities:
* `saveUserName` can be called multiple times, and it will change the current username of 
the user
* `deleteUserName` will delete the username of the user
Both endpoints can be called only by special/whitelisted smart contracts

### Impact:
* There is a new enable epoch definition for this feature, called `KeepExecOrderOnCreatedSCRsEnableEpoch`
* There are changes to the configuration options in the `systemSmartContractsConfig.toml` file, `[GovernanceSystemSCConfig]` section

> Also, the gas schedule files contain a new gas definition called `GetActiveFund`

Reviewer: I believe this should not be here if already mentioned above:
_Also, the gas schedule files contain a new gas definition called `GetActiveFund`_

Author: I would leave it here also as someone might just read the impact sections.

## 10. Balance data tries [#4636](https://github.com/multiversx/mx-chain-go/pull/4636)

The keys where the values are saved in a data trie are not random. Because of this, the data tries are not balanced, thus resulting in more intermediary nodes.
When saving some data in the data trie, do not save at `key` but rather at `hash(key)`. This way, the keys will be random, resulting in balanced data tries.
In order for this change to be backwards compatible, add Version field to trie nodes.
Nodes that are accessed will be automatically migrated to the new version (where the storage key is hash(key))
Added a builtin function that will migrate data trie nodes when it is called.
Added a new API endpoint which will return true if the data trie of a specified account is fully migrated.

### Impact:
* There is a new enable epoch definition for this feature, called `AutoBalanceDataTriesEnableEpoch`
* The gas schedule files are changed, there are 2 new gas costs added `TrieLoadPerNode` and `TrieStorePerNode`
* No binary flags changes
* There is one new API endpoint called `/address/:address/is-data-trie-migrated` that will return true if the account's data trie was migrated or false otherwise

> Relevant PRs:

Reviewer: add ### for _Relevant PRs_

Author: done

asyncCalls and callBacks are concluded.
scA registers a number of asyncCalls and finishes its execution. The VM iterates on the registered asyncCalls and calls them in the order they were created.
If one execution is intrashard, it will execute synchronously, in the case of cross-shard calls, those will be propagated as a smart contract result cross-shard,
and a callback is called through a callback SCR. The function which should be called back is registered and persisted in the storage, so the VM will read from
storage according to the unique IDs of the asyncCalls and get all the necessary information needed for the callback.
Introduced a limitation of multi-level asynchronous calls. AsyncCall from AsyncCall is not allowed. AsyncCall from callback is not allowed.

**ManagedBigFloats:**\
Introduced a new API that targets bigFloats. Safe math with bigFloats, using standardized GO libraries.

**ManagedMap:**\
Introduced a new API to support managed Maps for SC developers. Simple to use, all map features included.

**BackTransfers:**\

> We found that we can help developers, by changing some of the paradigms regarding SC to SC execution with payments. On the old VM if parentSC called childSC and childSC

Reviewer: ```suggestion
We found that we can help developers by changing some of the paradigms regarding SC to SC execution with payments. On the old VM if parentSC called childSC and childSC
```

Author: changed

* No binary flags changes
* No API endpoints changes

### Relevant PRs:
- [#5295](https://github.com/multiversx/mx-chain-go/pull/5295) - Removed baseAccount from peerAccount and added trackable data trie directly in user account
- [#5329](https://github.com/multiversx/mx-chain-go/pull/5329) - Moved the user and peer account in a separate package
- [#5361](https://github.com/multiversx/mx-chain-go/pull/5361) - Moved the data trie tracker in a separate package, and removed NewEmptyPeerAccount() function
- [#5366](https://github.com/multiversx/mx-chain-go/pull/5366) - Fixes after the refactoring work: changed an `if` condition, added unit tests

## 15. Full archive refactor [#5345](https://github.com/multiversx/mx-chain-go/pull/5345)

This feature is a complete refactoring for the full archive solution. The old solution was based on the idea that full archive nodes would still
connect on the p2p network by overriding the sharding counters but in practice, it did not perform well. This new solution relies on a secondary,
optional p2p network on which only the full archive nodes will join. Since this network will mostly contain full archive nodes, the connection

> between the nodes and the cycles request-response will be optimized.

Reviewer: ```suggestion
between the nodes and the request-response cycles will be optimized.
```

Author: done

### Impact:
* No activation epoch
* The `p2p.toml` file has been updated (removed the `[AdditionalConnections]` section) and the fullArchiveP2P.toml has been added
* There are 2 new optional binary flags added `full-archive-p2p-config` and `full-archive-port`
* No API endpoints changes

> Relevant PRs:

Reviewer: add ###

Author: done
